source/translated_table_model.ts

```diff
@@ -0,0 +1,210 @@
+  }
+
+  private sourceMove(operation: MoveRowOperation) {
+    const source = operation.source;
+    const destination = operation.destination;
+    const sourceIndex = this.sourceToTranslatedIndices[source];
+    const destinationIndex = this.sourceToTranslatedIndices[destination];
+    const originalIndices = {} as {[key: string]: number};
+    if(source > destination) {
+      for(let index = destination; index < source; index++) {
+        originalIndices[index] = this.sourceToTranslatedIndices[index];
+      }
+      for(let index = destination; index < source; index++) {
+        const translatedIndex = originalIndices[index];
+        this.translatedToSourceIndices[translatedIndex]++;
+        this.sourceToTranslatedIndices[
+          this.translatedToSourceIndices[translatedIndex]] = translatedIndex;
+      }
+    } else {
+      for(let index = source + 1; index <= destination; index++) {
+        originalIndices[index] = this.sourceToTranslatedIndices[index];
+      }
+      for(let index = source + 1; index <= destination; index++) {
+        const translatedIndex = originalIndices[index];
+        this.translatedToSourceIndices[translatedIndex]--;
+        this.sourceToTranslatedIndices[
+          this.translatedToSourceIndices[translatedIndex]] = translatedIndex;
+      }
+    }
+    this.translatedToSourceIndices[sourceIndex] = destination;
+    this.sourceToTranslatedIndices[destination] = sourceIndex;
+    this.moveRowHelper(sourceIndex, destinationIndex);
+  }
+
+  private sourceRemove(operation: RemoveRowOperation) {
+    const translatedIndex = this.sourceToTranslatedIndices[operation.index];
+    const rowCount = this.translatedToSourceIndices.length;
+    for(let index = 0; index < rowCount - 1; index++) {
+      if(index >= translatedIndex) {
+        this.translatedToSourceIndices[index] =
+          this.translatedToSourceIndices[index + 1];
+        this.sourceToTranslatedIndices[this.translatedToSourceIndices[index]] =
+          index;
+      }
+      const indexValue = this.translatedToSourceIndices[index];
+      if(indexValue >= operation.index) {
+        this.translatedToSourceIndices[index] = indexValue - 1;
+        this.sourceToTranslatedIndices[indexValue - 1] = index;
+      }
+    }
+    this.translatedToSourceIndices.pop();
+    this.sourceToTranslatedIndices.pop();
+    this.processOperation(new RemoveRowOperation(translatedIndex));
+  }
+
+  private sourceUpdate(operation: UpdateOperation) {
+    const translatedIndex = this.sourceToTranslatedIndices[operation.row];
+    this.processOperation(new UpdateOperation(translatedIndex,
+      operation.column));
+  }
+
+  private moveRowHelper = (source: number, destination: number) => {
+    const sourceValue = this.translatedToSourceIndices[source];
+    const iterations = Math.abs(source - destination);
+    const multiplier = source < destination ? 1 : -1;
+    for(let i = 0; i < iterations; i++) {
+      const updatedIndex = source + i * multiplier;
```

Comment:
Use the for loop's introducing statements to manage the index instead of decoupling them into an index and an updated index.

    for(let i = source; i != destination; i += multiplier)